README.md
#component-proxy-install

*Proxy git over http locally*

This is a hack to allow for private repos to be installed without exposing any
passwords on your local system. It will utilize ssh git checkouts on your
system in order to make it work. It only works with github.

##Using

Usage is almost identical to `component install` just with a different
invocation. In order to use run the following command::

    $ component-proxy-install

Any options you pass will be passed to `component install`

##How it works

It works by checking out all of your repos to a temporary directory
and serving those using a simple express app.

It serves the repos using the same interface as `raw.githubusercontent.com`


HISTORY.md
#component-proxy-install Changelog

##v0.4.1
* Removed version requirement check for component.
* Require the correct component.

##v0.4.0
* Switched to our own fork of component v0.18.0 called [component-x](https://github.com/wizcorp/component-x).

##v0.3.0
* Because of [annoying express send deprecation warnings](https://github.com/strongloop/express/issues/2227)
  we change our API usage and pin express version to >=4.7.0
* Converted README.rst to README.md
* Added this HISTORY.md


index.js
var crypto = require('crypto'),
	exec = require('child_process').exec,
	express = require('express'),
	fs = require('fs'),
	mime = require('mime'),
	ensureDir = require('mkdirp').sync,
	os = require('os'),
	path = require('path'),
	rimraf = require('rimraf'),
	util = require('util');

var CLONE_COMMAND = 'git clone git@github.com:%s/%s.git';
var CHECKOUT_COMMAND = 'git checkout %s';

var exists = fs.existsSync;

function isDir(path) {
	if(!exists(path)) {
		return false;
	}
	return fs.statSync(path).isDirectory();
}

function isFile(path) {
	if(!exists(path)) {
		return false;
	}
	return fs.statSync(path).isFile();
}

function newId() {
	return crypto.randomBytes(8).toString('hex');
}

var tmpDir = (typeof os.tmpdir === 'function') ? os.tmpdir() : os.tmpDir();
var componentsPath = path.resolve(tmpDir, newId());

/**
 * Clone the repository
 */
function cloneRepo(userDir, user, repo, tree, callback) {
	console.log('     cloning : %s/%s', user, repo);
	// Clone the repo or fail
	var command = util.format(CLONE_COMMAND, user, repo);

	exec(command, { cwd: userDir }, function(err, stdout, stderr) {
		if(err) {
			if(stderr.indexOf('Repository not found.') !== -1) {
				return callback('RepoNotFound');
			}
			return callback(err);
		}
		checkoutRepoTree(userDir, user, repo, tree, callback);
	});
}

/**
 * Checkout the repository's tree
 */
function checkoutRepoTree(userDir, user, repo, tree, callback) {
	var repoDir = path.resolve(userDir, repo);
	var command = util.format(CHECKOUT_COMMAND, tree);

	exec(command, { cwd: repoDir }, function(err, stdout, stderr) {
		if(err) {
			if(stderr.indexOf('did not match any file(s)') !== -1) {
				return callback('RepoNotFound');
			}
			return callback(err);
		}
		callback();
	});
}

/**
 * Serve a file from a repo
 */
function serveFile(fileToServe, res) {
	if(!isFile(fileToServe)) {
		return res.status(404).send('File not found');
	}

	res.type(mime.lookup(fileToServe));

	fs.readFile(fileToServe, function (err, data) {
		if (err) {
			return res.status(404).send('File not found');
		}

		res.status(200).send(data);
	});
}

var app = express();

app.get('/:user/:repo/:tree/*', function(req, res) {
	var user = req.params.user,
		repo = req.params.repo,
		tree = req.params.tree; // Branch/tag/commit

	var userDir = path.resolve(componentsPath, user),
		repoDir = path.resolve(userDir, repo);

	var repoKey = util.format('%s/%s', user, repo);
	var filePath = req.params[0];
	var fileToServe = path.join(userDir, repo, filePath);

	// Check if the repo exists
	if(isDir(repoDir)) {
		return serveFile(fileToServe, res);
	}

	// Create the user directory
	ensureDir(userDir);

	cloneRepo(userDir, user, repo, tree, function(err) {
		if(err) {
			if(err === 'RepoNotFound') {
				return res.status(404).send('File not found');
			}
			console.error(err);
			return res.status(500).send('Internal Server Error: ' + err);
		}
		// Serve the requested file

		serveFile(fileToServe, res);
	});
});

var server;

exports.start = function () {
	ensureDir(componentsPath);
	server = app.listen(0);
};

exports.getPort = function () {
	return server.address().port;
};

exports.stop = function (cb) {
	rimraf(componentsPath, function(error) {
		if (error) {
			console.warn('error while removing temp dir:', error);
		}

		server.close();
		cb();
	});
};


lib/app.js
var crypto = require('crypto'),
	exec = require('child_process').exec,
	express = require('express'),
	fs = require('fs'),
	mime = require('mime'),
	os = require('os'),
	path = require('path'),
	rimraf = require('rimraf'),
	util = require('util'),
	utils = require('./utils');

var CLONE_COMMAND = 'git clone git@github.com:%s/%s.git';
var UPDATE_COMMAND = 'git pull origin %s';
var CHECKOUT_COMMAND = 'git checkout %s';

function newId() {
	return crypto.randomBytes(8).toString('hex');
}

var componentsPath = path.resolve(os.tmpdir(), newId());

/**
 * Clone the repository
 */
function cloneRepo(userDir, user, repo, tree, callback) {
	console.log('     cloning : %s/%s', user, repo);
	// Clone the repo or fail
	var command = util.format(CLONE_COMMAND, user, repo);
	exec(command, { cwd: userDir }, function(err, stdout, stderr) {
		if(err) {
			if(stderr.indexOf('Repository not found.') !== -1) {
				var notFound = new Error('Repository not found');
				notFound.name = 'NotFound';
				return callback(notFound);
			}
			return callback(err);
		}
		checkoutRepoTree(userDir, user, repo, tree, callback);
	});
}

/**
 * Update the repository
 */
function updateRepo(userDir, user, repo, tree, callback) {
	console.log('    updating : %s/%s', user, repo);
	var command = util.format(UPDATE_COMMAND, tree);
	var repoDir = path.resolve(userDir, repo);
	exec(command, { cwd: repoDir }, function(err, stdout, stderr) {
		if(err) {
			return callback(err);
		}
		checkoutRepoTree(userDir, user, repo, tree, callback);
	});
}

/**
 * Checkout the repository's tree
 */
function checkoutRepoTree(userDir, user, repo, tree, callback) {
	var repoDir = path.resolve(userDir, repo);
	var command = util.format(CHECKOUT_COMMAND, tree);
	exec(command, { cwd: repoDir }, function(err, stdout, stderr) {
		if(err) {
			if(stderr.indexOf('did not match any file(s)') !== -1) {
				var notFound = new Error('Tree not found');
				notFound.name = 'NotFound';
				return callback(notFound);
			}
			return callback(err);
		}
		callback();
	});
}

/**
 * Serve a file from a repo
 */
function serveFile(fileToServe, res) {
	if(!utils.isFile(fileToServe)) {
		return res.send(404, 'File not found');
	}

	res.type(mime.lookup(fileToServe));

	fs.readFile(fileToServe, function (err, data) {
		if (err) {
			return res.send(404, 'File not found');
		}

		res.send(200, data);
	});
}

var app = express();

app.get('/:user/:repo/:tree/*', function(req, res) {
	var user = req.params.user,
		repo = req.params.repo,
		tree = req.params.tree; // Branch/tag/commit

	var userDir = path.resolve(componentsPath, user),
		repoDir = path.resolve(userDir, repo);

	var repoKey = util.format('%s/%s', user, repo);
	var filePath = req.params[0];
	var fileToServe = path.join(userDir, repo, filePath);

	// Check if the repo exists
	if(utils.isDir(repoDir)) {
		return serveFile(fileToServe, res);
	}

	// Create the user directory
	utils.ensureDir(userDir);

	cloneRepo(userDir, user, repo, tree, function(err) {
		if(err) {
			console.error(err);
			if(err.name == 'NotFound') {
				return res.send(404, 'Repo not found');
			}
			return res.send(500, 'Internal Server Error');
		}
		// Serve the requested file

		serveFile(fileToServe, res);
	});
});

var server;

exports.start = function () {
	utils.ensureDir(componentsPath);
	server = app.listen(0);
};

exports.getPort = function () {
	return server.address().port;
};

exports.stop = function (cb) {
	rimraf(componentsPath, function(error) {
		if (error) {
			console.warn('error while removing temp dir:', error);
		}

		server.close();
		cb();
	});
};
